twister: Add support for passing extra serial port to pytest

gchwier opened the pull request:
Extend twister's device serial handling to support multiple serial ports
for pytest-based tests. This enables testing scenarios that require
communication with multiple UART interfaces on the same device.
This enhancement enables comprehensive testing of multi-UART devices
while maintaining backward compatibility with single serial setups.

Updated from samples/sysbuild/hello_world (fixes problem from https://github.com/zephyrproject-rtos/zephyr/pull/95848, where interface identifiers are required and changed between if00 and if01 in test).

Tested with commands:
```
# using hardware map:
./scripts/twister -vv -ll debug --no-clean --west-flash=--erase -T samples/sysbuild/hello_world --device-testing --hardware-map ~/tmp/hardware_map.yml

# or passing second serial with command line:
./scripts/twister -vv -ll debug --no-clean --west-flash=--erase -T samples/sysbuild/hello_world --device-testing --device-serial /dev/ttyACM1 --device-serial /dev/ttyACM0 -p nrf54l15dk/nrf54l15/cpuapp

INFO    - 1/1 nrf54l15dk/nrf54l15/cpuapp sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr PASSED (device 7.911s <zephyr>)
INFO    -                                    sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr.test_uart_in_app    PASSED      
INFO    -                                    sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr.test_uart_in_second_core PASSED 

# when only one serial is given then second test is skipped:
./scripts/twister -vv -ll debug --no-clean --west-flash=--erase -T samples/sysbuild/hello_world --device-testing --device-serial /dev/ttyACM1 -p nrf54l15dk/nrf54l15/cpuapp

INFO    - 1/1 nrf54l15dk/nrf54l15/cpuapp sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr PASSED (device 3.777s <zephyr>)
INFO    -                                    sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr.test_uart_in_app    PASSED      
INFO    -                                    sample.sysbuild.hello_world.nrf54l15dk_nrf54l15_cpuflpr.test_uart_in_second_core SKIPPED      Second serial not available
```

gchwier committed:
scripts: twister: Add support for passing extra serial port to pytest

Extend twister's device serial handling to support multiple serial ports
for pytest-based tests. This enables testing scenarios that require
communication with multiple UART interfaces on the same device.
This enhancement enables comprehensive testing of multi-UART devices
while maintaining backward compatibility with single serial setups.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
samples: sysbuild: hello_world: Get second UART from Twister params

Use --device-serial command line option to specify multiple serial
ports. Split test into two separate test functions.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

JarmouniA reviewed (commented):
Can we have some documentation for this addition? Thanks!
Here for ex.
https://github.com/zephyrproject-rtos/zephyr/blob/433893e54b9ca7e70d6f158690839b56913d5dce/scripts/pylib/pytest-twister-harness/src/twister_harness/plugin.py#L63
Or/and here with more details
https://github.com/zephyrproject-rtos/zephyr/blob/433893e54b9ca7e70d6f158690839b56913d5dce/doc/develop/test/twister.rst?plain=1#L1259

JarmouniA reviewed (changes requested):
Missing documentation.

gchwier committed:
doc: twister: Document extra serial port

Update documentation for using multiple serial ports
with Twister's --devise-serial option.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

hakehuang commented on scripts/pylib/twister/twisterlib/hardwaremap.py:
```diff
@@ -195,7 +195,7 @@ def discover(self):
 
             elif self.options.device_serial:
-                self.add_device(self.options.device_serial,
+                self.add_device(self.options.device_serial[0],
```
would this un-compatible with former settings? I mean we used to set the serial in hwmap as string?

gchwier replied:
the first one:
```
- connected: true
  id: OLBZY05MJYIRV
  platform: mimxrt1060_evk@C/mimxrt1062/qspi
  product: linkserver
  runner: linkserver
  serial: /dev/ttyACM0
- connected: true
  id: "OLBZY05MJYIRV"
  platform: **unknown**
  product: Segger
  runner: jlink
  serial: /dev/ttyACM1
```
in second entry you should leave as was generated (`unknown`) or set the name of second core.
It is backward compatible.
